src/base/http/requestparser.cpp

```diff
@@ -30,314 +31,307 @@
+    }
 
-    request = m_request;
-    return NoError;
+    qWarning() << Q_FUNC_INFO << "unsupported request method: " << m_request.method;
+    return {ParseStatus::BadRequest, Request(), 0};  // TODO: SHOULD respond "501 Not Implemented"
 }
 
-bool RequestParser::parseStartingLine(const QString &line)
+bool RequestParser::parseStartLines(const QString &data)
 {
-    const QRegExp rx("^([A-Z]+)\\s+(\\S+)\\s+HTTP/\\d\\.\\d$");
-
-    if (rx.indexIn(line.trimmed()) >= 0) {
-        m_request.method = rx.cap(1);
-
-        QUrl url = QUrl::fromEncoded(rx.cap(2).toLatin1());
-        m_request.path = url.path(); // Path
-
-        // Parse GET parameters
-        QListIterator<QPair<QString, QString> > i(QUrlQuery(url).queryItems());
-        while (i.hasNext()) {
-            QPair<QString, QString> pair = i.next();
-            m_request.gets[pair.first] = pair.second;
+    // for simplicity, we don't handle malformed request which uses `LF` for newline
+    const QVector<QStringRef> lines = data.splitRef(CRLF, QString::SkipEmptyParts);
+
+    // [rfc7230] 3.2.2. Field Order
+    QStringList requestLines;
+    for (const auto &line : lines) {
+        if (line.at(0).isSpace() && !requestLines.isEmpty()) {
+            // continuation of previous line
+            requestLines.last() += line.toString();
+        }
+        else {
+            requestLines += line.toString();
         }
-
-        return true;
     }
 
-    qWarning() << Q_FUNC_INFO << "invalid http header:" << line;
-    return false;
-}
+    if (requestLines.isEmpty())
+        return false;
 
-bool RequestParser::parseHeaderLine(const QString &line, QPair<QString, QString> &out)
-{
-    int i = line.indexOf(QLatin1Char(':'));
-    if (i == -1) {
-        qWarning() << Q_FUNC_INFO << "invalid http header:" << line;
+    if (!parseRequestLine(requestLines[0]))
         return false;
+
+    for (auto i = ++(requestLines.begin()); i != requestLines.end(); ++i) {
+        if (!parseHeaderLine(*i, m_request.headers))
+            return false;
     }
 
-    out = qMakePair(line.left(i).trimmed().toLower(), line.mid(i + 1).trimmed());
     return true;
 }
 
-bool RequestParser::parseHttpHeader(const QByteArray &data)
+bool RequestParser::parseRequestLine(const QString &line)
 {
-    QString str = QString::fromUtf8(data);
-    QStringList lines = str.trimmed().split(EOL);
-
-    QStringList headerLines;
-    foreach (const QString &line, lines) {
-        if (line[0].isSpace()) { // header line continuation
-            if (!headerLines.isEmpty()) { // really continuation
-                headerLines.last() += QLatin1Char(' ');
-                headerLines.last() += line.trimmed();
-            }
-        }
-        else {
-            headerLines.append(line);
-        }
-    }
+    // [rfc7230] 3.1.1. Request Line
 
-    if (headerLines.isEmpty())
-        return false; // Empty header
+    const QRegularExpression re(QLatin1String("^([A-Z]+)\\s+(\\S+)\\s+HTTP\\/(\\d\\.\\d)$"));
```

Comment:
Request line has very simple format. Isn't using regular expressions too wasteful to parse it?

Reply:
>Isn't using regular expressions too wasteful to parse it?

Maybe, but I don't want to fight with all kinds of invalid inputs, IMO if the input can match this regex then we take it.